src/samplers/resampler.jl

```diff
@@ -0,0 +1,53 @@
+"""
+    WeightedResampler(obs::AbstractArray, wv::AbstractWeights)
+
+A WeightedResampler is a subtype of Distributions.Sampleable which randomly selects
+observations from the raw input data (`obs`) based on the weights (`wv`) provided.
+
+This type supports univariate, multivariate and matrixvariate forms, so `obs` can
+be a vector of values, matrix of values or a vector of matrices.
+"""
+struct WeightedResampler{F<:VariateForm, S<:ValueSupport, T<:AbstractArray} <: Sampleable{F, S}
+    obs::T
+    wv::AbstractWeights
+end
+
+function WeightedResampler(obs::T, wv::AbstractWeights) where T<:AbstractArray
```

Comment:
Is there any reason to restrict weights to `AbstractWeights` rather than `AbstractVector`? At https://github.com/JuliaLang/julia/pull/31395, I made functions accept any array since there's no ambiguity.

Reply:
ping @rofinn @nickrobinson251

Reply:
Idk but maybe @rofinn does

But also it'd be an easy follow up to loosen it

Reply:
ping @rofinn ?

Reply:
Can we merge as is and open an issue about this? It'd be a non breaking change to loosen the constraint in future. Or just loosen it now and fix it if there's a bug report

Reply:
I guess merging this as-is is OK. I've noticed that the weights are passed directly to `sample`, so if we loosen the signature we should also ensure we convert the argument to `AbstractWeights`, which is what `sample` expects (currently at least). But I still think loosening this is a good idea, though not a requirement.